bindings/matrix-sdk-ffi/src/authentication_service.rs

```diff
@@ -121,6 +125,56 @@ impl AuthenticationService {
         }
     }
 
+    /// Restore an existing session on the current homeserver using an access
+    /// token issued by an authentication server.
+    /// # Arguments
+    ///
+    /// * `token` - The access token issued by the authentication server.
+    ///
+    /// * `device_id` - The device ID that the access token was scoped for.
+    pub fn restore_with_access_token(
+        &self,
+        token: String,
+        device_id: String,
+    ) -> Result<Arc<Client>, AuthenticationError> {
+        match self.client.read().as_ref() {
+            Some(client) => {
+                // Restore the client and ask the server for the full user ID as this
+                // could be different from the username that was entered.
+                let discovery_user_id = UserId::parse("@unknown:unknown")
+                    .map_err(|e| AuthenticationError::Generic { message: e.to_string() })?;
+                let device_id: OwnedDeviceId = device_id.as_str().into();
+
+                let discovery_session = Session {
+                    access_token: token.clone(),
+                    user_id: discovery_user_id,
+                    device_id: device_id.clone(),
+                };
+
+                client.restore_session(discovery_session).map_err(AuthenticationError::from)?;
+                let whoami = client.whoami()?;
+
+                // Create the actual client with a store path from the user ID.
+                let homeserver_url = client.homeserver();
+                let session =
+                    Session { access_token: token, user_id: whoami.user_id.clone(), device_id };
+                let client = Arc::new(ClientBuilder::new())
+                    .base_path(self.base_path.clone())
+                    .homeserver_url(homeserver_url)
+                    .username(whoami.user_id.to_string())
+                    .build()
+                    .map_err(AuthenticationError::from)?;
+
+                // Restore the client using the session.
+                client
```

Comment:
Do you want to modify the original client behind the mutex as well, or is that not important?

Reply:
Hmmm that's an interesting point. I don't see it as important as the service should be finished with when this succeeds. I think its fine for now but maybe we'll need to revisit this in the future 👍